nameko/events.py

```diff
@@ -183,10 +168,63 @@ def __init__(self, source_service, event_type, handler_type=SERVICE_POOL,
         super(EventHandler, self).__init__(
             queue=None, requeue_on_error=requeue_on_error)
 
+    @property
+    def broadcast_identifier(self):
+        """ A unique string to identify a service instance for `BROADCAST`
+        type handlers.
+
+        The `broadcast_identifier` is appended to the queue name when the
+        `BROADCAST` handler type is used. It must uniquely identify service
+        instances that receive broadcasts.
+
+        The default `broadcast_identifier` is a uuid that is set when the
+        service starts. It will change when the service restarts, meaning
+        that any unconsumed messages that were broadcast to the 'old' service
+        instance will not be received by the 'new' one. ::
+
+            @property
+            def broadcast_identifier(self):
+                # use a uuid as the identifier.
+                # the identifier will change when the service restarts and
+                # any unconsumed messages will be lost
+                return uuid.uuid4().hex
+
+        The default behaviour is therefore incompatible with reliable delivery.
+
+        An alternative `broadcast_identifier` that would survive service
+        restarts is ::
+
+            @property
+            def broadcast_identifier(self):
+                # use the machine hostname as the identifier.
+                # this assumes that only one instance of a service runs on
+                # any given machine
+                return socket.gethostname()
+
+        If neither of these approaches are appropriate, you could read the
+        value out of a configuration file ::
+
+            @property
+            def broadcast_identifier(self):
+                return self.config['SERVICE_IDENTIFIER']  # or similar
+
+        Broadcast queues are exclusive to ensure that `broadcast_identifier`
+        values are unique.
+        """
+        if self.reliable_delivery and self.handler_type is BROADCAST:
```

Comment:
As far as I can see, the only control you get over `getmembers` is the `predicate` argument, which evaluates each item. That would invoke the descriptors if you're inspecting an instance of a class.

For the `bind` case specifically we can avoid the problem by inspecting the class rather than the instance, but we also call `getmembers` in `iter_extensions` and that isn't a simple change. I would really like to remove `iter_extensions` and daisy-chain calls to `setup`, `start`, `stop` etc. by making extensions responsible for looking after their subextensions, but that is a bigger issue for another time. I guess the summary is that's not an insurmountable problem.

In the mean time, "is safe" is nebulous but there is at least a concrete description: the descriptor on an extension will be called during container creation (after the extension's `__init__`, before `setup`) and may be called at any point afterwards.

As for healthy APIs... I completely agree that they should be driven by concepts not code reuse. I guess I like the `handler_type` part of the existing API - it makes it clear that the value changes how the handler behaves, as opposed to `broadcast_event`, which might suggest that the event type is different. That's a discussion for another time though.

What would you like me to do about this PR? I'm still fairly keen on using a descriptor with a caveat in the docs and an intention to improve it later.